Improve box model article (#15538)
mdn/content: files/en-us/learn/css/building_blocks/the_box_model/index.md

## Block and inline boxes

> In CSS we broadly have two types of boxes — **block boxes** and **inline boxes**. The type affects the behavior of the box in terms of page flow and in relation to other boxes on the page. Boxes have an **inner display type** and an **outer display type**. Let's start with the outer type.

Reviewer: ```suggestion
In CSS we broadly have two types of boxes — **block boxes** and **inline boxes**. The type refers to how the box behaves in terms of page flow and in relation to other boxes on the page. Boxes have an **inner display type** and an **outer display type**. First, we will explain what we mean by block box and inline box. We will then explain what is meant by an inner and outer display type.
```

Reviewer: original better describes the flow of the tutorial (we aren't starting with outer type)

> You can set the values of the {{cssxref("display")}} property to `block` or `inline`. This defines the type of box applied to an element and relates to the outer display type.

Reviewer: there are many other values of display

Author: I wanted to make clear what the display type is in CSS - the first time the display: x property is displayed in the article is line 80 - for flex (inner). Then, in line 91 for block. Maybe I can move it up and rephrase, see my next commit.

Author: See https://github.com/mdn/content/pull/15538/commits/3b164aa5d67c9b6119475aa2cfe275b6cf26741f

## Block and inline boxes

In CSS we broadly have two types of boxes — **block boxes** and **inline boxes**. The type refers to how the box behaves in terms of page flow and in relation to other boxes on the page. Boxes have an **inner display type** and an **outer display type**.

In general, you can set various values for the display type using the {{cssxref("display")}} property, which can have various values.

## Outer display type

If a box has an outer display type of `block`, then:

- The box will break onto a new line.
- The {{cssxref("width")}} and {{cssxref("height")}} properties are respected.
- Padding, margin and border will cause other elements to be pushed away from the box.

> - The box will extend in the inline direction to fill the space available in its container. In most cases the box will become as wide as its container, filling up 100% of the space available.

Reviewer: ```suggestion
- The box will extend in the inline direction to fill the space available in its container. In most cases, the box will become as wide as its container, filling up 100% of the space available.
```

## Block and inline boxes

> In CSS we broadly have two types of boxes — **block boxes** and **inline boxes**. The type affects the behavior of the box in relation to page flow and other boxes on the page. Boxes have an **inner display type** and an **outer display type**. We'll start explaining the block box and inline box concept. And later, we'll tackle the inner and outer display types. 

Author: Not sure why 'broadly' was used.... and, outer display type used in line 61 is unknown to the reader....

```suggestion
In CSS the `display` property can define boxes of type block and inline. These block types affect the behavior of the box in relation to page flow and other boxes on the page. This is the outer display type.
```

Author: I suggest to change the text in line 61 and below as well - reason: The user know the outer display type but nothing else:

```
## The inner display type

As mentioned above, boxes in CSS have an _outer_ display type. Boxes also have an _inner_ display type, which dictates how elements inside that box are laid out. By default, the elements inside a box are laid out in **[normal flow](/en-US/docs/Learn/CSS/CSS_layout/Normal_Flow)**, which means that they behave just like any other block and inline elements.
We can change the inner display type by using `display` values like `flex`. The outer display type remains `block`. Any direct children of this box will become flex items and will be laid out according to the rules set out in the [Flexbox](/en-US/docs/Learn/CSS/CSS_layout/Flexbox) spec, which you'll learn about later on.
> **Note:** To read more about how boxes work in block and inline layout, take a look at the MDN guide to [Block and Inline Layout](/en-US/docs/Web/CSS/CSS_Flow_Layout/Block_and_Inline_Layout_in_Normal_Flow).
```

> In CSS we broadly have two types of boxes — **block boxes** and **inline boxes**. These characteristics refer to how the box behaves in terms of page flow and in relation to other boxes on the page. Boxes also have an **inner display type** and an **outer display type**. First, we will explain what we mean by block box and inline box. We will then explain what is meant by an inner and outer display type.

Author: another idea

Author: ```suggestion
In CSS we broadly have two types of boxes — **block boxes** and **inline boxes**. The type affects the behavior of the box in relation to page flow and other boxes on the page. Boxes have an **inner display type** and an **outer display type**. We'll start explaining the block box and inline box concept. And later, we'll tackle the inner and outer display type. 
```

Some HTML elements, such as `<a>`, `<span>`, `<em>` and `<strong>` use `inline` as their outer display type by default.

> ## Aside: The inner display type

Reviewer: If we decide to break the page up in this new way, we need to add ## outer display type heading above (line #51)  and change this to ## inner display type (no "aside")

> You've learned above that boxes in CSS have an _outer_ display type. Boxes also have an _inner_ display type, which dictates how elements inside that box are laid out.

Reviewer: ```suggestion
Boxes also have an _inner_ display type, which dictates how elements inside that box are laid out.
```

Block and inline layout is the default way things behave on the web. By default and without any other instruction, the elements inside a box are also laid out in **[normal flow](/en-US/docs/Learn/CSS/CSS_layout/Normal_Flow)** and behave as block or inline boxes.

You can change the inner display type for example by setting `display: flex;`. The element will still use the outer display type `block` but this changes the inner display type to `flex`. Any direct children of this box will become flex items and behave according to the [Flexbox](/en-US/docs/Learn/CSS/CSS_layout/Flexbox) spec.

If you look more into CSS layout, you will also come across other inner values your boxes can have, for example [`grid`](/en-US/docs/Learn/CSS/CSS_layout/Grids).

> **Note:** To read more about the values of display, and how boxes work in block and inline layout, take a look at the MDN guide to [Block and Inline Layout](/en-US/docs/Web/CSS/CSS_Flow_Layout/Block_and_Inline_Layout_in_Normal_Flow).  

## Examples of different display types

> Below we have three different HTML elements, all of which have an outer display type of `block`.

Author: ```suggestion
The example below has three different HTML elements, all of which have an outer display type of `block`. 
```

{{EmbedGHLiveSample("css-examples/learn/box-model/inline.html", '100%', 1100)}}

The key thing to remember for now is: Changing the value of the `display` property can change whether the outer display type of a box is block or inline. This changes the way it displays alongside other elements in the layout.

## What is the CSS box model?

The CSS box model as a whole applies to block boxes and defines how the different parts of a box — margin, border, padding, and content — work together to create a box that you can see on a page. Inline boxes use just _some_ of the behavior defined in the box model.

To add complexity, there is a standard and an alternate box model. By default, browsers use the standard box model.

### Parts of a box

Making up a block box in CSS we have the:

- **Content box**: The area where your content is displayed; size it using properties like {{cssxref("width")}} and {{cssxref("height")}}.
- **Padding box**: The padding sits around the content as white space; size it using {{cssxref("padding")}} and related properties.
- **Border box**: The border box wraps the content and any padding; size it using {{cssxref("border")}} and related properties.

> - **Margin box**: The margin is the outermost layer, wrapping the content, padding, and border as whitespace between this box and other elements, size it using {{cssxref("margin")}} and related properties.

Author: ```suggestion
- **Margin box**: The margin is the outermost layer, wrapping the content, padding, and border as whitespace between this box and other elements; size it using {{cssxref("margin")}} and related properties.
```

Some HTML elements, such as `<a>`, `<span>`, `<em>` and `<strong>` use `inline` as their outer display type by default.

## Inner display type

Boxes also have an _inner_ display type, which dictates how elements inside that box are laid out.

Block and inline layout is the default way things behave on the web. By default and without any other instruction, the elements inside a box are also laid out in **[normal flow](/en-US/docs/Learn/CSS/CSS_layout/Normal_Flow)** and behave as block or inline boxes.

> You can change the inner display type for example by setting `display: flex;`. The element will still use the outer display type `block` but this changes the inner display type to `flex`. Any direct children of this box will become flex items and behave according to the [Flexbox](/en-US/docs/Learn/CSS/CSS_layout/Flexbox) spec.

Reviewer: ```suggestion
You can change the inner display type for example by setting `display: flex;`. The element will still use the outer display type `block` but this changes the inner display type to `flex`. Any direct children of this box will become flex items and behave according to the [Flexbox](/en-US/docs/Learn/CSS/CSS_layout/Flexbox) specification.
```

If you look more into CSS layout, you will also come across other inner values your boxes can have, for example [`grid`](/en-US/docs/Learn/CSS/CSS_layout/Grids).

> > **Note:** To read more about the values of display, and how boxes work in block and inline layout, take a look at the MDN guide to [Block and Inline Layout](/en-US/docs/Web/CSS/CSS_Flow_Layout/Block_and_Inline_Layout_in_Normal_Flow).  

Reviewer: ```suggestion
> **Note:** To read more about the values of display, and how boxes work in block and inline layout, take a look at the MDN guide [Block and Inline Layout](/en-US/docs/Web/CSS/CSS_Flow_Layout/Block_and_Inline_Layout_in_Normal_Flow).  
```

## Examples of different display types

The example below has three different HTML elements, all of which have an outer display type of `block`.

> - A paragraph with a border added in CSS. The browser renders this as a block box. The paragraph starts on a new line and uses the entire available width.

Reviewer: ```suggestion
- A paragraph with a border added in CSS. The browser renders this as a block box. The paragraph starts on a new line and extends the entire available width.
```

> - A list, which is laid out using `display: flex`. This establishes flex layout for the items inside the container. The list itself is a block box and — like the paragraph — expands to the full container width and breaks onto a new line.

Reviewer: ```suggestion
- A list, which is laid out using `display: flex`. This establishes flex layout for the children of the container, which are flex items. The list itself is a block box and — like the paragraph — expands to the full container width and breaks onto a new line.
```

> - A block-level paragraph, inside which are two `<span>` elements. These elements would normally be `inline`, however, one of the elements has a class of block, and we have set it to `display: block`.

Reviewer: ```suggestion
- A block-level paragraph, inside which are two `<span>` elements. These elements would normally be `inline`, however, one of the elements has a class of "block" which gets set to `display: block`.
```

{{EmbedGHLiveSample("css-examples/learn/box-model/block.html", '100%', 1100)}}

In the next example, we can see how `inline` elements behave.

- The `<span>` elements in the first paragraph are inline by default and so do not force line breaks.

> - The `<ul>` element that is set to `display: inline-flex` creates an inline box around some flex items.

Reviewer: ```suggestion
- The `<ul>` element that is set to `display: inline-flex` creates an inline box containing some flex items.
```

- The two paragraphs are both set to `display: inline`. The inline flex container and paragraphs all run together on one line rather than breaking onto new lines (as they would do if they were displaying as block-level elements).

**To toggle between the display modes, you can change `display: inline` to `display: block` or `display: inline-flex` to `display: flex`.**

{{EmbedGHLiveSample("css-examples/learn/box-model/inline.html", '100%', 1100)}}

The key thing to remember for now is: Changing the value of the `display` property can change whether the outer display type of a box is block or inline. This changes the way it displays alongside other elements in the layout.

## What is the CSS box model?

The CSS box model as a whole applies to block boxes and defines how the different parts of a box — margin, border, padding, and content — work together to create a box that you can see on a page. Inline boxes use just _some_ of the behavior defined in the box model.

To add complexity, there is a standard and an alternate box model. By default, browsers use the standard box model.

### Parts of a box

Making up a block box in CSS we have the:

> - **Content box**: The area where your content is displayed; size it using properties like {{cssxref("width")}} and {{cssxref("height")}}.

Reviewer: ```suggestion
- **Content box**: The area where your content is displayed; size it using properties like  {{cssxref("inline-size")}} and {{cssxref("block-size")}} or {{cssxref("width")}} and {{cssxref("height")}}.
```

- **Padding box**: The padding sits around the content as white space; size it using {{cssxref("padding")}} and related properties.
- **Border box**: The border box wraps the content and any padding; size it using {{cssxref("border")}} and related properties.
- **Margin box**: The margin is the outermost layer, wrapping the content, padding, and border as whitespace between this box and other elements; size it using {{cssxref("margin")}} and related properties.

The below diagram shows these layers:

![Diagram of the box model](box-model.png)

### The standard CSS box model

> In the standard box model, if you give a box a `width` and a `height` attribute, this defines the width and height of the _content box_. Any padding and border is then added to that width and height to get the total size taken up by the box (see image below).

Reviewer: ```suggestion
In the standard box model, if you give a box an `inline-size` and a `block-size` (or `width` and a `height`) attributes, this defines the inline-size and block-size (with and height in horizontal languages) of the _content box_. Any padding and border is then added to those dimensions to get the total size taken up by the box (see image below).
```

### The alternative CSS box model

In the alternative box model, any width is the width of the visible box on the page. The content area width is that width minus the width for the padding and border (see image below). No need to add up the border and padding to get the real size of the box.

> To turn on the alternative model for an element, set `box-sizing: border-box` on it.

Author: ```suggestion
To turn on the alternative model for an element, set `box-sizing: border-box` on it:
```

```css
.box {
  box-sizing: border-box;
}
```

If we assume the box has the same CSS as above:

```css
.box {
  width: 350px;
  height: 150px;
  margin: 10px;
  padding: 25px;
  border: 5px solid black;
}
```

> The _actual_ space taken up by the box will be 350px wide and 150px high.

Author: ```suggestion
Now, the _actual_ space taken up by the box will be 350px wide and 150px high.
```

> If we assume the box has the same CSS as above:
> 
> ```css
> .box {
>   width: 350px;
>   height: 150px;
>   margin: 10px;
>   padding: 25px;
>   border: 5px solid black;
> }
> ```
> 
> Now, the _actual_ space taken up by the box will be 350px wide and 150px high.

Author: Just in case we do not want to repeat the example

```suggestion
If we assume the box has the same values as the CSS in the example above, the _actual_ space taken up by the box is now 350px wide and 150px high.
```

### The alternative CSS box model

In the alternative box model, any width is the width of the visible box on the page. The content area width is that width minus the width for the padding and border (see image below). No need to add up the border and padding to get the real size of the box.

To turn on the alternative model for an element, set `box-sizing: border-box` on it:

```css
.box {
  box-sizing: border-box;
}
```

If we assume the box has the same CSS as above:

```css
.box {

>   width: 350px;

Reviewer: ```suggestion
  width: 350px;
  inline-size: 350px;
```

>   height: 150px;

Reviewer: ```suggestion
  height: 150px;
  block-size: 150px;
```

  margin: 10px;
  padding: 25px;
  border: 5px solid black;
}
```

> Now, the _actual_ space taken up by the box will be 350px wide and 150px high.

Reviewer: ```suggestion
Now, the _actual_ space taken up by the box will be 350px in the inline direction and 150px in the block direction.
```

}
```

> If you want to understand the thinking behind this, see [the CSS Tricks article on box-sizing](https://css-tricks.com/inheriting-box-sizing-probably-slightly-better-best-practice/).

Author: ```suggestion
To understand the underlying idea, you can read [the CSS Tricks article on box-sizing](https://css-tricks.com/inheriting-box-sizing-probably-slightly-better-best-practice/).in
```

#### Margin collapsing

A key thing to understand about margins is the concept of **margin collapsing**:
- Positive margins on two elements whose margins touch will combine to become one margin. Its size will be equal to the largest individual margin.
- If one margin is negative, its value will be _subtracted_ from the total.

> - Where both are negative, the margins will collapse and the smallest (furthest from zero) value will be used.

Author: I don't like this list yet :/

Author: ```suggestion
Depending on whether margins are positive or negative on two elements whose margins touch, the results will be different:
- Two positive margins will combine to become one margin. Its size will be equal to the largest individual margin.
- Two negative margins will collapse and the smallest (furthest from zero) value will be used.
- If one margin is negative, its value will be _subtracted_ from the total.
```

### Padding

The padding sits between the border and the content area and is used to push the content away from the border. Unlike margins, you cannot have a negative padding. Any background applied to your element will display behind the padding.

The {{cssxref("padding")}} property controls the padding on all sides of an element. To control each side individually, use these longhand properties:

- {{cssxref("padding-top")}}
- {{cssxref("padding-right")}}
- {{cssxref("padding-bottom")}}
- {{cssxref("padding-left")}}

In the example below, you can change the values for padding on the class `.box` to see that this changes where the text begins in relation to the box. You can also change the padding on the class `.container` to create space between the container and the box. You can change the padding on any element to create space between its border and whatever is inside the element.

{{EmbedGHLiveSample("css-examples/learn/box-model/padding.html", '100%', 700)}}

## The box model and inline boxes

All of the above fully applies to block boxes. Some of the properties can apply to inline boxes too, such as those created by a `<span>` element.

> In the example below, we have a `<span>` inside a paragraph and have applied a `width`, `height`, `margin`, `border`, and `padding` to it. You can see that the width and height are ignored. The vertical margin, padding, and border are respected but they do not change the relationship of other content to our inline box and so the padding and border overlaps other words in the paragraph. Horizontal padding, margins, and borders are respected and will cause other content to move away from the box.

Author: ```suggestion
In the example below, we have a `<span>` inside a paragraph. We have applied a `width`, `height`, `margin`, `border`, and `padding` to it. You can see that the width and height are ignored. The vertical margin, padding, and border are respected but don't change the relationship of other content to our inline box. The padding and border overlap other words in the paragraph. The horizontal padding, margins, and borders move other content away from the box.
```
